File: backend/quantum_simulation.py
"""
Quantum simulation utilities for multi-qubit systems
Main entry point that re-exports from modular components
"""

try:
    # Try relative imports first (when imported as module)
    from .circuit_operations import (
        QuantumGate,
        QuantumCircuit,
        get_gate_matrix,
        apply_gate,
        simulate_circuit,
        create_initial_state,
        partial_trace,
        calculate_bloch_vector,
        EXAMPLE_CIRCUITS
    )
except ImportError:
    # Fall back to absolute imports (when run directly)
    from circuit_operations import (
        QuantumGate,
        QuantumCircuit,
        get_gate_matrix,
        apply_gate,
        simulate_circuit,
        create_initial_state,
        partial_trace,
        calculate_bloch_vector,
        EXAMPLE_CIRCUITS
    )

# Matrix operations
def matrix_multiply(a, b):
    """Matrix multiplication"""
    import numpy as np
    return np.dot(a, b)

def tensor_product(a, b):
    """Tensor product of two matrices"""
    import numpy as np
    return np.kron(a, b)

def transpose(matrix):
    """Matrix transpose"""
    import numpy as np
    return matrix.T

def trace(matrix):
    """Matrix trace"""
    import numpy as np
    return np.trace(matrix)

# Legacy exports for backward compatibility
def matrix_trace(matrix):
    return trace(matrix)

File: src/utils/quantum/quantumSimulation.ts
// Quantum simulation utilities for multi-qubit systems
// Main entry point that re-exports from modular components

export { GATES, PAULI, AVAILABLE_GATES, SINGLE_QUBIT_GATES, TWO_QUBIT_GATES, THREE_QUBIT_GATES, getGateMatrix, validateGateParameters } from './gates';
export { matrixMultiply, tensorProduct, trace, transpose, matrixAdd, scalarMultiply, identity, zeros, isSquare, isHermitian, frobeniusNorm, matrixEquals } from '../core/matrixOperations';
export { createInitialState, calculateBlochVector, calculateSuperposition, calculateEntanglement, calculateConcurrence, calculateVonNeumannEntropy, calculateReducedDensityMatrix, calculateEntanglementWitness, formatDensityMatrix, partialTrace } from './densityMatrix';
export { applyGate, simulateCircuit, computeGateOutputState, simulateCircuitWithStates, EXAMPLE_CIRCUITS, testGateOutputs } from './circuitOperations';

// Re-export types and interfaces
export type { QuantumGate } from './gates';
export type { DensityMatrix } from './densityMatrix';
export type { QuantumCircuit } from './circuitOperations';

// Legacy exports for backward compatibility
import { trace } from '../core/matrixOperations';

export const matrixTrace = (matrix: number[][]): number => {
  return trace(matrix);
};
